netbox_branching/merge_strategies/squash.py

```diff
@@ -0,0 +1,790 @@
+        collapsed.depended_by = set()
+
+    # Preemptively detect and split bidirectional FK cycles
+    # This may add new UPDATE operations to to_process
+    _split_bidirectional_cycles(to_process, logger)
+
+    # Group by action and sort each group by time - need this to build the
+    # dependency graph correctly
+    deletes = sorted(
+        [v for v in to_process.values() if v.final_action == 'delete'],
+        key=lambda c: c.last_change.time
+    )
+    updates = sorted(
+        [v for v in to_process.values() if v.final_action == 'update'],
+        key=lambda c: c.last_change.time
+    )
+    creates = sorted(
+        [v for v in to_process.values() if v.final_action == 'create'],
+        key=lambda c: c.last_change.time
+    )
+
+    _build_fk_dependency_graph(deletes, updates, creates, logger)
+    ordered_keys = _dependency_order_by_references(to_process, logger)
+
+    # Convert keys back to collapsed changes
+    result = [to_process[key] for key in ordered_keys]
+    return result
+
+
+class SquashMergeStrategy(MergeStrategy):
+    """
+    Squash merge strategy that collapses multiple changes per object into a single operation.
+    """
+
+    def merge(self, branch, changes, request, commit, logger):
+        """
+        Apply changes after collapsing them by object and ordering by dependencies.
+        """
+        models = set()
+
+        logger.info("Collapsing ObjectChanges by object...")
+        collapsed_changes = {}
+
+        for change in changes:
+            key = (change.changed_object_type.id, change.changed_object_id)
+
+            if key not in collapsed_changes:
+                model_class = change.changed_object_type.model_class()
+                collapsed = CollapsedChange(key, model_class)
+                collapsed_changes[key] = collapsed
+                logger.debug(f"New object: {model_class.__name__}:{change.changed_object_id}")
+
+            collapsed_changes[key].changes.append(change)
+
+        # Collapse each object's changes
+        for key, collapsed in collapsed_changes.items():
+            collapsed.collapse(logger)
+
+        # Order collapsed changes based on dependencies
+        ordered_changes = _order_collapsed_changes(collapsed_changes, logger)
+
+        # Apply collapsed changes in order
+        logger.info(f"Applying {len(ordered_changes)} collapsed changes...")
+        for i, collapsed in enumerate(ordered_changes, 1):
+            model_class = collapsed.model_class
+            models.add(model_class)
+
+            last_change = collapsed.last_change
+
+            logger.info(f"  [{i}/{len(ordered_changes)}] {collapsed.final_action.upper()} "
+                       f"{model_class.__name__}:{collapsed.key[1]} "
+                       f"(from {len(collapsed.changes)} original changes)")
+
+            with event_tracking(request):
+                request.id = last_change.request_id
+                request.user = last_change.user
```

Comment:
Because we're potentially compressing multiple changes that may come from different users, I'm not sure it makes sense to attribute them all to the user who made the most recent change. I would argue that wherever two or more changes are being squashed, the resulting change should be attributed to the user who is merging the branch, because that is the only person we should assume has full contextual awareness.

This topic porbably deserves further discussion outside the context of this PR; I just wanted to flag it here for further consideration.